### DISCUSSION TITLE
Please provide a convenient solution for Mac owners to use the nightly build of muCommander

### DISCUSSION THREAD

**Discussion by ghost** [mucommander/mucommander#716] [Ideas]
Please provide a convenient solution for Mac owners to use the nightly build of muCommander.

The currently existing solution is probably not acceptable for most Mac owners, for the reasons given in #695.

**Comment by ahadas** [comment]
@Andreas0602 the nightly build is mostly for testing and for that running it with the `mucommander.sh` script is simple enough.
Let's convert this to a discussion and talk about your use case since I believe it's pretty unique

**Comment by ahadas** [comment]
@Andreas0602 how about the following procedure:
1. Clone the repo mucommander/mucommander (including the submodules)
2. Create a dmg with `./gradlew dmg`
3. Install the dmg
4. Replace the jars in /Applications with the ones from the nightly build

Step 4 is basically what you used to do before, the problem is that this doesn't work anymore with the signed dmg so in step 1 you'll create an unsigned jar as we had in the past. Could you give this a try?

**Comment by ghost** [comment]
@ahadas, thank you. The procedure that you proposed does work.

Some questions/comments:

- Before cloning the repo and creating the dmg, I had to install JDK, Xcode (to get git), and create and upload a public SSH key. I guess not every Mac owner is willing to do that.
- I don't understand why step 4 is necessary. Isn't the master branch of the repo identical to the nightly build?
- Is there any disadvantage in using the unsigned dmg created in step 2 compared to using a signed dmg? (I only made a few, quick tests so far and couldn't find any disadvantage.)

**Reply by ahadas** [reply]
> @ahadas, thank you. The procedure that you proposed does work.

Great, happy to hear :)

> 
> Some questions/comments:
> 
> * Before cloning the repo and creating the dmg, I had to install JDK, Xcode (to get git), and create and upload a public SSH key.


Right, JDK is needed in order to compile the project, but you don't have to install xcode - git can be installed [using brew](https://formulae.brew.sh/formula/git)

> I guess not every Mac owner is willing to do that.

True, but the nightly build is not for every user - users that want to experience with new things that get in to the master branch or help with testing, need to have git and JDK

> * I don't understand why step 4 is necessary. Isn't the master branch of the repo identical to the nightly build?

Ah yeah, of course - step 4 is not needed when the nightly build is the same as the master branch at the time of creating the dmg. But let's say that tomorrow there's a new nightly build with some changes, then you can update your installed app that way (copying the jars) without creating and installing a dmg (in other words, once you install an unsigned dmg then you can only do step 4 as long as there are no changes to other scripts or resources within the app)

> * Is there any disadvantage in using the unsigned dmg created in step 2 compared to using a signed dmg? (I only made a few, quick tests so far and couldn't find any disadvantage.)

From a functionality point of view, there's no difference.
The benefit in having a signed dmg is that users that download it (i.e., don't build it locally) can easily install it without workarounds for passing macOS's gatekeeper

**Comment by pskowronek** [comment]
My 2 cents - what about alpha release lines? Now muC informs that a new release version is present, and exactly the same could be for alpha if one opted for it. Such releases could have log levels increased etc. I assume nightly could be troublesome as it would require CI/CD pipeline that includes signing DMGs etc.

Such alpha releases (or even maybe official ones?) could have "submit a bug" button or menu item where muC would pre-fill crucial information about muC version, system, java ver etc (probably could be done with: https://docs.github.com/en/enterprise-server@3.2/issues/tracking-your-work-with-issues/creating-an-issue#creating-an-issue-from-a-url-query ?) 

This way muC devs could have feedback about their changes sooner than later (like after release).

**Reply by ahadas** [reply]
> My 2 cents - what about alpha release lines? Now muC informs that a new release version is present, and exactly the same could be for alpha if one opted for it.

it's possible to have a similar mechanism for alpha/nightly versions - that was the intention of having `nightly.xml` and `version.xml` configurations but in practice we've noticed that users didn't really use nightly versions. sometimes, when it took long to release a new version, users tried the nightly build to test specific functionality, then switched back to the 'stable' release and waited for that functionality to be included in the next 'stable' release. I know that @Andreas0602 uses the nightly build now but I don't think this has changed significantly, especially when we release more often (every 2-3 months lately) to justifying packaging alpha releases (note that it should be simple to switch back to the previous version, which is still available, if needed)

> Such releases could have log levels increased etc. I assume nightly could be troublesome as it would require CI/CD pipeline that includes signing DMGs etc.

right and besides signing, the credentials for connecting to Google Drive and Dropbox needs to be set - otherwise that version won't include all functionality. even just creating a dmg / rpm / deb requires building the release on the appropriate operating system and that wasn't simple (I currently have a virtual machine with jenkins that creates and publishes the nightly build on Fedora) - this part can be simplified maybe using GitHub actions, which is on my todo list to check

> Such alpha releases (or even maybe official ones?) could have "submit a bug" button or menu item where muC would pre-fill crucial information about muC version, system, java ver etc (probably could be done with: https://docs.github.com/en/enterprise-server@3.2/issues/tracking-your-work-with-issues/creating-an-issue#creating-an-issue-from-a-url-query ?)

that's interesting - it's hard to estimate how much we'll gain from that compared to the link to GitHub issues we currently have. one would anyway need to sign in to GitHub I suppose.. and if so, the different between doing it via the browser or the application is not that significant, no?

> This way muC devs could have feedback about their changes sooner than later (like after release).

I think the portable version is ok for testing basic functionality. I agree that if we're able to publish release candidates that include all functionality and are packaged the same way as the 'stable' releases, it would give us more confidence before releasing a 'stable' release but with the current process of releasing a version, I'm not sure it is worth the trouble
